### PR TITLE
Add show/hide button to password fields

### PR DESCRIPTION
**Pull request by SleeplessOne1917**
## Description
Gives the user the ability to toggle the visibility of their password when using password input fields.

[Screencast_20230707_191946.webm](https://github.com/LemmyNet/lemmy-ui/assets/28871516/afb38546-a573-4bc1-a545-ae40c8d3fea4)

### With forgot password link
image

### With password strengthometer
image

I also tweaked some parts of the settings page.

## Before
image
image

## After
image
image

### CODE REVIEW

File: src/shared/components/home/login.tsx [hunk 1]
```diff
@@ -90,28 +90,14 @@ export class Login extends Component<any, State> {
...
-              >
-                {I18NextService.i18n.t("forgot_password")}
-              </NavLink>
-            </div>
+          <div className="mb-3">
+            <PasswordInput
+              id="login-password"
+              value={this.state.form.password}
+              onInput={linkEvent(this, this.handleLoginPasswordChange)}
+              label={I18NextService.i18n.t("password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/home/login.tsx:98]: Should probably remove that `?? undefined`

File: src/shared/components/home/setup.tsx [hunk 2]
```diff
@@ -121,41 +122,21 @@ export class Setup extends Component<any, State> {
...
-              minLength={10}
-              maxLength={60}
-            />
-          </div>
+        <div className="mb-3">
+          <PasswordInput
+            id="password"
+            value={this.state.form.password}
+            onInput={linkEvent(this, this.handleRegisterPasswordChange)}
+            label={I18NextService.i18n.t("password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/home/setup.tsx:130]: Same.

File: src/shared/components/home/signup.tsx [hunk 3]
```diff
@@ -210,57 +182,26 @@ export class Signup extends Component<any, State> {
...
-              </div>
-            )}
-          </div>
+        <div className="mb-3">
+          <PasswordInput
+            id="register-password"
+            value={this.state.form.password}
+            onInput={linkEvent(this, this.handleRegisterPasswordChange)}
+            showStrength
+            label={I18NextService.i18n.t("password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/home/signup.tsx:191]: Same.

File: src/shared/components/home/signup.tsx [hunk 4]
```diff
@@ -210,57 +182,26 @@ export class Signup extends Component<any, State> {
...
-              className="form-control"
-              required
-            />
-          </div>
+        <div className="mb-3">
+          <PasswordInput
+            id="register-verify-password"
+            value={this.state.form.password_verify}
+            onInput={linkEvent(this, this.handleRegisterPasswordVerifyChange)}
+            label={I18NextService.i18n.t("verify_password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/home/signup.tsx:200]: Same.

File: src/shared/components/person/password-change.tsx [hunk 5]
```diff
@@ -60,37 +61,22 @@ export class PasswordChange extends Component<any, State> {
...
-              maxLength={60}
-            />
-          </div>
+        <div className="mb-3">
+          <PasswordInput
+            id="new-password"
+            value={this.state.form.password}
+            onInput={linkEvent(this, this.handlePasswordChange)}
+            showStrength
+            label={I18NextService.i18n.t("new_password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/person/password-change.tsx:70]: Same.

File: src/shared/components/person/password-change.tsx [hunk 6]
```diff
@@ -60,37 +61,22 @@ export class PasswordChange extends Component<any, State> {
...
-              required
-              maxLength={60}
-            />
-          </div>
+        <div className="mb-3">
+          <PasswordInput
+            id="password"
+            value={this.state.form.password_verify}
+            onInput={linkEvent(this, this.handleVerifyPasswordChange)}
+            label={I18NextService.i18n.t("verify_password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/person/password-change.tsx:78]: Same.

File: src/shared/components/person/settings.tsx [hunk 7]
```diff
@@ -318,59 +319,30 @@ export class Settings extends Component<any, SettingsState> {
...
-                onInput={linkEvent(this, this.handleNewPasswordChange)}
-              />
-            </div>
+          <div className="mb-3">
+            <PasswordInput
+              id="new-password"
+              value={this.state.changePasswordForm.new_password}
+              onInput={linkEvent(this, this.handleNewPasswordChange)}
+              showStrength
+              label={I18NextService.i18n.t("new_password") ?? undefined}
```
Comment (dessalines) [inline src/shared/components/person/settings.tsx:328]: Same.

File: src/shared/components/common/password-input.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  {
+    id: 3,
+    value: "strong",
+    minDiversity: 4,
+    minLength: 14,
+  },
+];
+
+function handleToggleShow(i: PasswordInput) {
+  i.setState(prev => ({
```
[the remark this answers is not in the document]
Comment (SleeplessOne1917) [reply src/shared/components/common/password-input.tsx:51]: If inferno is anything like react, setting state that depends on previous state should be done with functions.

[the remark this answers is not in the document]
Comment (SleeplessOne1917) [reply src/shared/components/common/password-input.tsx:51]: [Here is an article going into what I'm talking about](https://medium.com/@wisecobbler/using-a-function-in-setstate-instead-of-an-object-1f5cfd6e55d1).

File: src/shared/components/common/password-input.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+import { I18NextService } from "../../services";
+import { Icon } from "./icon";
+
+interface PasswordInputProps {
+  id: string;
+  value?: string;
+  onInput: FormEventHandler<HTMLInputElement>;
+  className?: string;
+  showStrength?: boolean;
+  label?: string | null;
```
Comment (dessalines) [inline src/shared/components/common/password-input.tsx:15]: If you have `?`, I don't think `| null` is necessary.

File: src/shared/components/common/password-input.tsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                required
+                maxLength={60}
+                minLength={showStrength ? 10 : 0}
+              />
+              <button
+                className="btn btn-outline-dark"
+                type="button"
+                id={id}
+                onClick={linkEvent(this, handleToggleShow)}
+                aria-label={show ? "Hide Password" : "Show Password"}
```
[the remark this answers is not in the document]
Comment (dessalines) [reply src/shared/components/common/password-input.tsx:106]: Some of the strings in this file still need fixed.

**Review by dessalines — COMMENTED** [review]
Ok translations are merged now.

File: src/shared/components/common/password-input.tsx (hunk 9 above)
Comment (SleeplessOne1917) [reply src/shared/components/common/password-input.tsx:15]: The `i18n.t` function returns a null if the translation doesn't exist. That's why there were so many null coalesces before.

File: src/shared/components/common/password-input.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (SleeplessOne1917) [reply src/shared/components/common/password-input.tsx:106]: Which ones?

**Review by dessalines — APPROVED** [review]

**Pull request merged**